src/server/worker/api_server.go

```diff
@@ -1778,107 +1773,112 @@ func (a *APIServer) worker() {
 				// previously-created job has finished, or job was finished during backoff
 				// or in the 'watcher' queue
 				logger.Logf("skipping job %v as it is already in state %v", jobID, jobPtr.State)
-				continue NextJob
+				continue
 			}
 
 			// create new ctx for this job, and don't use retryCtx as the
 			// parent. Just because another job's etcd write failed doesn't
 			// mean this job shouldn't run
-			jobCtx, jobCancel := context.WithCancel(a.pachClient.Ctx())
-			defer jobCancel() // cancel the job ctx
-			pachClient := a.pachClient.WithCtx(jobCtx)
-
-			//  Watch for any changes to EtcdJobInfo corresponding to jobID; if
-			// the EtcdJobInfo is marked 'FAILED', call jobCancel().
-			// ('watcher' above can't detect job state changes--it's watching
-			// an index and so only emits when jobs are created or deleted).
-			go a.cancelCtxIfJobFails(jobCtx, jobCancel, jobID)
-
-			// Inspect the job and make sure it's relevant, as this worker may be old
-			jobInfo, err := pachClient.InspectJob(jobID, false)
-			if err != nil {
-				if col.IsErrNotFound(err) {
-					continue NextJob // job was deleted--no sense retrying
+			if err := func() error {
```

Comment:
Another big problem it caught, the `defer jobCancel` below would never run, which meant we were leaking a goroutine here.